sysl2/sysl/deps.go

```diff
@@ -0,0 +1,264 @@
+			next := MakeAppElement(nextAppName, nextEpName)
+			nextDep := MakeAppDependency(target, next, stmt)
+			walkPassthrough(excludes, passthroughs, nextDep, integrations, module)
+		}
+	}
+}
+
+func (ds *DependencySet) FindIntegrations(apps, excludes, passthroughs StrSet, module *sysl.Module) *DependencySet {
+	integrations := NewDependencySet()
+	outboundDeps := NewDependencySet()
+	lenPassthroughs := len(passthroughs)
+	for _, dep := range ds.Deps {
+		appNames := dep.extractAppNames()
+		endpoints := dep.extractEndpoints()
+		isSubsection := isSub(appNames, apps)
+		isSelfSubsection := isSub(MakeStrSet(dep.Self.Name), apps)
+		isTargetSubsection := isSub(MakeStrSet(dep.Target.Name), passthroughs)
+		interExcludes := appNames.Intersection(excludes)
+		interEndpoints := endpoints.Intersection(MakeStrSet(".. * <- *", "*"))
+		lenInterExcludes := len(interExcludes)
+		lenInterEndpoints := len(interEndpoints)
+		if isSubsection && lenInterExcludes == 0 && lenInterEndpoints == 0 {
+			integrations.Deps[dep.String()] = dep
+		}
+		// collect outbound dependencies
+		if lenPassthroughs > 0 &&
+			((isSubsection || (isSelfSubsection && isTargetSubsection)) && lenInterExcludes == 0 && lenInterEndpoints == 0) {
+			outboundDeps.Deps[dep.String()] = dep
+		}
+	}
+	if lenPassthroughs > 0 {
+		for _, dep := range outboundDeps.Deps {
+			walkPassthrough(excludes, passthroughs, dep, integrations, module)
+		}
+	}
+
+	return integrations
+}
+
+func (ds *DependencySet) CollectAppDependencies(module *sysl.Module) {
+	for appname, app := range module.GetApps() {
+		for epname, endpoint := range app.GetEndpoints() {
+			ds.collectStatementDependencies(endpoint.GetStmt(), appname, epname)
+		}
+	}
+}
+
+func (ds *DependencySet) collectStatementDependencies(stmts []*sysl.Statement, appname, epname string) {
+	for _, stat := range stmts {
+		switch c := stat.GetStmt().(type) {
+		case *sysl.Statement_Call:
+			targetName := getAppName(c.Call.GetTarget())
+			dep := MakeAppDependency(MakeAppElement(appname, epname), MakeAppElement(targetName, c.Call.GetEndpoint()), stat)
+			ds.Deps[dep.String()] = dep
+		case *sysl.Statement_Action, *sysl.Statement_Ret:
+			continue
+		case *sysl.Statement_Cond:
+			ds.collectStatementDependencies(c.Cond.GetStmt(), appname, epname)
+		case *sysl.Statement_Loop:
+			ds.collectStatementDependencies(c.Loop.GetStmt(), appname, epname)
+		case *sysl.Statement_LoopN:
+			ds.collectStatementDependencies(c.LoopN.GetStmt(), appname, epname)
+		case *sysl.Statement_Foreach:
+			ds.collectStatementDependencies(c.Foreach.GetStmt(), appname, epname)
+		case *sysl.Statement_Group:
+			ds.collectStatementDependencies(c.Group.GetStmt(), appname, epname)
+		case *sysl.Statement_Alt:
+			for _, choice := range c.Alt.GetChoice() {
+				ds.collectStatementDependencies(choice.GetStmt(), appname, epname)
+			}
+		default:
+			panic("No statement!")
+		}
+	}
+}
+
+func isSub(child, parent StrSet) bool {
+	if len(child) == 0 && len(parent) == 0 {
```

Comment:
what is `isSub`? Please give it a better name.

Reply:
changed it to IsSubSet and move it to StrSet.go file